File: Testcases/GoSource/samples/CodeUnderTest.go
package main

import (
	"errors"
	"math/big"
	"math/rand"
	"time"
)

// a set of functions to be tested

// Echo my input

func EchoInt(in int) (out int) {
	randomSleep(50 * time.Millisecond)
	out = in
	return
}

func EchoFloat(in float64) (out float64) {
	randomSleep(50 * time.Millisecond)
	out = in
	return
}

func EchoString(in string) (out string) {
	randomSleep(50 * time.Millisecond)
	out = in
	return
}

// Sum my inputs


func SumInt(in1, in2 int) (out int) {
	randomSleep(50 * time.Millisecond)
	out = in1 + in2
	return
}

func SumFloat(in1, in2 float64) (out float64) {
	randomSleep(5)
	out = in1 + in2
	return
}

func SumString(in1, in2 string) (out string) {
	randomSleep(50 * time.Millisecond)
	out = in1 + in2
	return
}

// Factorial computation: factorial(n):
// n < 0 - undefined
// n == 0 - 1
// n > 0 - n * factorial(n-1)

var ErrInvalidInput = errors.New("invalid input")

// Factorial via iteration
func FactorialIterate(n int64) (res *big.Int, err error) {
	if n < 0 {
		err = ErrInvalidInput
		return
	}
	res = big.NewInt(1)
	if n == 0 {
		return
	}
	for  i := int64(1); i <= n; i++ {
		term := big.NewInt(i)
		//res.Mul(res, big.NewInt(i))
		res = term.Mul(term, res)
	}
	return
}

// Factorial via recursion
func FactorialRecurse(n int64) (res *big.Int, err error) {
	if n < 0 {
		err = ErrInvalidInput
		return
	}
	res = big.NewInt(1)
	if n == 0 {
		return
	}
	term := big.NewInt(n)
	facm1, err := FactorialRecurse(n - 1)
	if err != nil {
		return
	}
	res = term.Mul(term, facm1)
	return
}

// a helper

func randomSleep(dur time.Duration ) {
	time.Sleep(time.Duration((1 + rand.Intn(int(dur)))))
}

File: Testcases/JavaSource/org/baf/CodeUnderTest.java
package org.baf;

import java.math.BigInteger;
import java.util.Random;

public class CodeUnderTest {

  public static int echoInt(int in) {
    randomSleep(50);
    return in;
  }

  public static double echoFloat(double in) {
    randomSleep(50);
    return in;
  }

  public static String echoString(String in) {
    randomSleep(50);
    return in;
  }

  public static int sumInt(int in1, int in2) {
    randomSleep(50);
    return in1 + in2;
  }

  public static double sumFloat(double in1, double in2) {
    randomSleep(50);
    return in1 + in2;
  }

  public static String sumString(String in1, String in2) {
    randomSleep(50);
    return in1 + in2;
  }

//Factorial computation: factorial(n):
//n < 0 - undefined
//n == 0 - 1
//n > 0 - n * factorial(n-1)

  public static BigInteger factorialIterative(long n) {
    if (n < 0) {
      throw new IllegalArgumentException("invalid input");
    }
    BigInteger res = BigInteger.ONE;
    if (n == 0) {
      return res;
    }
    for (long i = 1; i <= n; i++) {
      res = res.multiply(new BigInteger(Long.toString(i)));
    }
    return res;
  }

  public static BigInteger factorialRecursive(long n) {
    if (n < 0) {
      throw new IllegalArgumentException("invalid input");
    }
    BigInteger res = BigInteger.ONE;
    if (n == 0) {
      return res;
    }
    // return new BigInteger(Long.toString(n)).multiply(factorialRecursive(n - 1));
    return factorialRecursive(n - 1).multiply(new BigInteger(Long.toString(n)));
  }

  private static Random rand = new Random();

  private static void randomSleep(int durMs) {
    try {
      Thread.sleep(1 + rand.nextInt(durMs));
    } catch (InterruptedException e) {
      // ignore
    }

  }

}
